models/multiphase/d2q9_pseudo_potential/Dynamics.c.Rt

```diff
@@ -0,0 +1,209 @@
+    f[2] = uf;
+    uf = f[7];
+    f[7] = f[5];
+    f[5] = uf;
+    uf = f[8];
+    f[8] = f[6];
+    f[6] = uf;
+}
+
+CudaDeviceFunction void CollisionBGK() {
+    // Here we perform a single relaxation time collision operation.
+    // We save memory here by using a single dummy variable
+    real_t u_eq[2], d, f_temp[9];
+    d = getRho();
+    
+    vector_t u_eq_vect = getUeq();
+    u_eq[0] = u_eq_vect.x; 
+    u_eq[1] = u_eq_vect.y;
+
+    f_temp[0] = f[0];
+    f_temp[1] = f[1];
+    f_temp[2] = f[2];
+    f_temp[3] = f[3];
+    f_temp[4] = f[4];
+    f_temp[5] = f[5];
+    f_temp[6] = f[6];
+    f_temp[7] = f[7];
+    f_temp[8] = f[8];
+    SetEquilibrium(d, u_eq); //stores equilibrium distribution in f[0]-f[8]
+    f[0] = f_temp[0] - omega*(f_temp[0]-f[0]);	
+    f[1] = f_temp[1] - omega*(f_temp[1]-f[1]);
+    f[2] = f_temp[2] - omega*(f_temp[2]-f[2]);
+    f[3] = f_temp[3] - omega*(f_temp[3]-f[3]);	
+    f[4] = f_temp[4] - omega*(f_temp[4]-f[4]);
+    f[5] = f_temp[5] - omega*(f_temp[5]-f[5]);
+    f[6] = f_temp[6] - omega*(f_temp[6]-f[6]);	
+    f[7] = f_temp[7] - omega*(f_temp[7]-f[7]);
+    f[8] = f_temp[8] - omega*(f_temp[8]-f[8]);
+}
+
+CudaDeviceFunction real_t getRho() {
+    // This function defines the macroscopic density at the current node.
+    return f[8]+f[7]+f[6]+f[5]+f[4]+f[3]+f[2]+f[1]+f[0];
+}
+
+CudaDeviceFunction vector_t getU() {
+    // This function defines the macroscopic velocity at the current node.
+    // eq 6.28 from The Lattice Boltzmann Method: Principles and Practice
+    // T. Krüger, H. Kusumaatmaja, A. Kuzmin, O. Shardt, G. Silva, E.M. Viggen
+    real_t d = getRho();
+    vector_t u;
+
+    vector_t F_ff = getF_ff();
+    vector_t F_sf = getF_sf();
+
+    u.x = ( f[8]-f[7]-f[6]+f[5]-f[3]+f[1] )/d + (GravitationX + F_ff.x + F_sf.x )*0.5 ;
+    u.y = (-f[8]-f[7]+f[6]+f[5]-f[4]+f[2] )/d + (GravitationY + F_ff.y + F_sf.y )*0.5 ;
+    u.z = 0;
+    return u;
+}
+
+CudaDeviceFunction vector_t getUeq() {
+    // Schan-Chen forcing scheme 
+    // eq 6.29 from The Lattice Boltzmann Method: Principles and Practice
+    // T. Krüger, H. Kusumaatmaja, A. Kuzmin, O. Shardt, G. Silva, E.M. Viggen
+    real_t d = getRho();
+    vector_t F_ff = getF_ff();
+    vector_t F_sf = getF_sf();
+    vector_t u_eq;
+
+    u_eq.x = ( f[8]-f[7]-f[6]+f[5]-f[3]+f[1] )/d;
+    u_eq.y = (-f[8]-f[7]+f[6]+f[5]-f[4]+f[2] )/d;
+
+    u_eq.x += (GravitationX + (F_ff.x + F_sf.x)/d) / omega;
+    u_eq.y += (GravitationY + (F_ff.y + F_sf.y)/d) / omega;
+    u_eq.z = 0;
+
+    return u_eq;
+}
+
+    
+// ------------------------------ pseudo-potential related code --------------------------------------------- //
+
+CudaDeviceFunction void calcPsi() {
+    // Calculate psi at each point so that pseudopotential force can be found.
+    // eq 9.102 p369 from book: 'The Lattice Boltzmann Method: Principles and Practice'
+    // T. Krüger, H. Kusumaatmaja, A. Kuzmin, O. Shardt, G. Silva, E.M. Viggen
+    
+    real_t rho = getRho();
```

Comment:
This is might be ineffective. Create global variable, set it once and reuse. How many times/collision only you call getRho?